src/main/java/fi/aalto/cs/apluscourses/intellij/model/CourseUpdater.java

```diff
@@ -0,0 +1,147 @@
+package fi.aalto.cs.apluscourses.intellij.model;
+
+import com.intellij.notification.Notification;
+import com.intellij.openapi.application.ReadAction;
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.intellij.notifications.NewModulesVersionsNotification;
+import fi.aalto.cs.apluscourses.intellij.notifications.Notifier;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.Module;
+import fi.aalto.cs.apluscourses.utils.Event;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class CourseUpdater {
+
+  @NotNull
+  private final Course course;
+
+  @NotNull
+  private final Project project;
+
+  @NotNull
+  private final URL courseUrl;
+
+  @NotNull
+  private final Event eventToTrigger;
+
+  @NotNull
+  private final Notifier notifier;
+
+  private final long updateInterval;
+
+  private Thread thread = null;
+
+  private final Set<String> notifiedModules = new HashSet<>();
+
+  private Notification moduleUpdatesNotification = null;
+
+  /**
+   * Construct a course updater with the given parameters.
+   */
+  public CourseUpdater(@NotNull Course course,
+                       @NotNull Project project,
+                       @NotNull URL courseUrl,
+                       @NotNull Event eventToTrigger,
+                       @NotNull Notifier notifier,
+                       long updateInterval) {
+    this.course = course;
+    this.project = project;
+    this.courseUrl = courseUrl;
+    this.eventToTrigger = eventToTrigger;
+    this.notifier = notifier;
+    this.updateInterval = updateInterval;
+  }
+
+  /**
+   * Starts or restarts this updater. This method is thread safe and can be called from multiple
+   * threads.
+   */
+  public synchronized void restart() {
+    if (thread != null) {
+      thread.interrupt();
+    }
+    thread = new Thread(this::run);
+    thread.start();
+  }
+
+  /*
+   * The run lock ensures that only one thread executes the 'run' method. The run lock guarantees
+   * that a new thread created by 'restart' blocks until the previous thread has received the
+   * interrupt and exited. Note, that making the 'run' method 'synchronized' would cause a deadlock,
+   * because 'restart' is synchronized as well.
+   */
+  private final Object runLock = new Object();
+
+  private void run() {
+    synchronized (runLock) {
+      while (true) { //  NOSONAR
+        try {
+          var newCourse = fetchCourse();
```

Comment:
Nit:  it's a bit misleading that we create a `Course` object only to copy data from it to our original `Course` object, isn't it? Would it be possible that we wouldn't construct a new object but only the fetched modules as a list or something?

Reply:
Constructing an `IntelliJCourse` object is also only reason that this class must be in `intellij` package, right?

Reply:
Btw, would it be reasonable to check if the thread is interrupted after the fetching is done?

Reply:
> Btw, would it be reasonable to check if the thread is interrupted after the fetching is done?
Yes, very good idea, the quicker it responds to interrupts the better

Reply:
> Nit: it's a bit misleading that we create a Course object only to copy data from it to our original Course object, isn't it? Would it be possible that we wouldn't construct a new object but only the fetched modules as a list or something?

Yeah true, this is a pretty lazy way of reusing the existing parsing mechanism :smile: I'll either make the relevant module JSON parsing method from `Course` public or just write some JSON parsing here, and only deal with a list of modules instead of constructing a course

Reply:
> Constructing an IntelliJCourse object is also only reason that this class must be in intellij package, right?

Also the notifications make it part of `intellij` (although I'm sure there are some who would want to add a layer of abstraction over notifications to separate more code from the platform :wink:)

Reply:
Also now realized that not constructing an `IntelliJCourse` would also conveniently avoid the read lock, which is always nice